### PR TITLE
2994 new eccd current drive model to account for zeff and te with a cutoff function

### PR DESCRIPTION
**Pull request by chris-ashe**
## Description

New model has been input to account for ECCD cutoff as a function of electron temperature and effective charge

- [x] Dependance of the harmonic number has been verified
- [x] Dependance between O and X mode has been verified
- [x] Quantified the difference between `effrfss` & `effcd`
- [x] ECRH HARE model seems to be removed. Should tidy up
- [x] Update plotting dependencies eg `plot_proc`.

## Checklist

I confirm that I have completed the following checks:

- [x] I have justified any large differences in the regression tests caused by this pull request in the comments.
- [x] I have added new tests where appropriate for the changes I have made.
- [x] If I have had to change any existing unit or integration tests, I have justified this change in the pull request comments.
- [x] If I have made documentation changes, I have checked they render correctly.
- [x] I have added documentation for my change, if appropriate.

### CODE REVIEW

File: process/current_drive.py [hunk 1]
```diff
@@ -226,7 +286,7 @@ def cudriv(self, output: bool):
                     * 1.0e6
                 )
                 faccdfix = auxiliary_cdfix / physics_variables.plascur
-            elif current_drive_variables.iefrffix in [3, 7, 10, 11, 12]:
```
Comment (timothy-nunn) [inline process/current_drive.py:229]: Why was `11` removed here? Was it a mistake?

File: process/current_drive.py [hunk 2]
```diff
@@ -403,11 +463,100 @@ def cudriv(self, output: bool):
...
+                )
+
+                density_factor = 0.5e0 * (
+                    1.0e0 + np.tanh((2.0e0 / a) * ((fp - fc) / fp - a))
+                )
+
+                current_drive_variables.effcd = (
+                    current_drive_variables.effcd * density_factor
+                )
+                effrfss = effrfss * density_factor
```
Comment (timothy-nunn) [inline process/current_drive.py:497]: 470 to this line doesn't make sense to me. Why have we changed how the old switch behaves in this PR?

File: process/current_drive.py [hunk 3]
```diff
@@ -721,13 +880,27 @@ def cudriv(self, output: bool):
                 "(gamma_ecrh)",
                 current_drive_variables.gamma_ecrh,
             )
-        elif current_drive_variables.iefrf == 12:
+        if current_drive_variables.iefrf == 12:
```
Comment (timothy-nunn) [inline process/current_drive.py:883]: why?

File: process/current_drive.py [hunk 4]
```diff
@@ -721,13 +880,27 @@ def cudriv(self, output: bool):
...
                 current_drive_variables.xi_ebw,
             )
+        if current_drive_variables.iefrf in [12, 13]:
+            po.ovarre(
+                self.outfile,
+                "EC harmonic number",
+                "(harnum)",
+                current_drive_variables.harnum,
+            )
+        if current_drive_variables.iefrf == 13:
```
Comment (timothy-nunn) [inline process/current_drive.py:897]: Would more 'correct' to move this block under the `elif current_drive_variables.iefrf == 12:` and have it as `elif current_drive_variables.iefrf == 13:` and then have the if in block below as is.

**Review by timothy-nunn — CHANGES_REQUESTED** [review]
Couple of questions that need addressing.

File: process/current_drive.py (hunk 1 above)
Comment (chris-ashe) [reply process/current_drive.py:229]: The HARE model that we used to have had a switch value of 11. The model was removed but its switch remnants still remained in some places. Was tidying up

File: process/current_drive.py (hunk 2 above)
Comment (chris-ashe) [reply process/current_drive.py:497]: Realised that the cutoff function section was commented out so added it back in. The cutoff function was in for the `iefrffix` but not `iefrf`

File: process/current_drive.py (hunk 1 above)
Comment (timothy-nunn) [reply process/current_drive.py:229]: Ah ok. Could you note in the docs that 11 has been removed?

**Comment by timothy-nunn** [comment]
Also @chris-ashe, if you have some reference values would it be possible to add in a quick test for this

File: process/current_drive.py (hunk 1 above)
Comment (chris-ashe) [reply process/current_drive.py:229]: I am updating all of the current drive docs in https://github.com/ukaea/PROCESS/issues/2998. Which captures this change

File: process/current_drive.py (hunk 4 above)
Comment (chris-ashe) [reply process/current_drive.py:897]: It doesn't really matter in the order of the output. At the moment the output goes efficiency scaling value -> harmonic number -> wave mode. `iefrf = 13` dose not a efficiency scaling value.

**Comment by chris-ashe** [comment]
Test is now covered in #3013

**Review by jonmaddock — APPROVED** [review]
Looks great after rebasing.

**Comment by jonmaddock** [comment]
As Tim's questions have been answered and Chris is keen to get this in, I hope you don't mind me merging this Tim!

**Pull request merged**